web3/middleware/stalecheck.py

```diff
@@ -0,0 +1,55 @@
+import datetime
+import time
+
+
+SKIP_STALECHECK_FOR_METHODS = set([
+    'eth_getBlockByHash',
+    'eth_getBlockByNumber',
+    'eth_getBlockTransactionCountByHash',
+    'eth_getBlockTransactionCountByNumber',
+    'eth_getTransactionByBlockHashAndIndex',
+    'eth_getTransactionByBlockNumberAndIndex',
+    'eth_getUncleCountByBlockHash',
+    'eth_getUncleCountByBlockNumber',
+])
+
+
+def _isfresh(block, allowable_delay):
+    return block and time.time() - block.timestamp <= allowable_delay
+
+
+def make_stalecheck_middleware(**kwargs):
+    '''
+    Use to require that a function will run only of the blockchain is recently updated.
+
+    This middleware takes an argument, so unlike other middleware, you must make the middleware
+    with a method call.
+    For example: `make_stalecheck_middleware(hours=12)`
+
+    If the latest block in the chain is older than 12 hours ago in this example, then the
+    middleware will raise a StaleBlockchain exception.
+
+    `make_stalecheck_middleware(...)` takes the same keyword arguments as datetime.timedelta()
+    '''
+    allowable_delay = datetime.timedelta(**kwargs).total_seconds()
```

Comment:
On a second look, I do see that this is an ugly use of `kwargs`. For supporting a `datetime.datetime`, I think the added complexity outweighs the value. I'm having trouble imagining when that would be the best way to specify the allowable delay.

Here is an attempt to succinctly and clearly cover the first and third point (in spirit):

`def make_stalecheck_middleware(seconds=0, minutes=0, hours=0, days=0):`

What do you think?

Reply:
I still think it's a cleaner API to allow any of the three because they serve the most common use cases.

* I would use seconds in cases where my allowable delay is small (an hour or less)
* I would use `datetime.timedelta` or seconds when my delay gets bigger.
* I would use `datetime.datetime` if I knew the exact moment where I needed to be able to read the data (this is the weakest use case).

If you don't want to do negotiation of the three types and you want a simple API, I'd recommend just leaving it as seconds only and making people do the conversion outside of this function.